File: chillnote/Core/Components/NoteSourceCard.swift
import SwiftUI

struct NoteSourceCard: View {
    let source: NoteSourceMetadata
    var compact: Bool = false

    @Environment(\.openURL) private var openURL

    var body: some View {
        Button {
            guard let url = URL(string: source.url) else { return }
            openURL(url)
        } label: {
            HStack(spacing: 10) {
                sourceBadge

                VStack(alignment: .leading, spacing: compact ? 1 : 2) {
                    Text(source.platformName)
                        .font(.chillCaption.weight(.semibold))
                        .foregroundColor(.textSub)
                        .lineLimit(1)

                    Text(source.title)
                        .font(compact ? .bodySmall : .bodyMedium)
                        .foregroundColor(.textMain)
                        .lineLimit(compact ? 1 : 2)
                        .multilineTextAlignment(.leading)
                }

                Spacer(minLength: 8)

                Image(systemName: "arrow.up.right")
                    .font(.system(size: compact ? 12 : 13, weight: .semibold))
                    .foregroundColor(.textSub)
            }
            .padding(.horizontal, compact ? 10 : 12)
            .padding(.vertical, compact ? 9 : 11)
            .background(Color.bgSecondary)
            .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
            .contentShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
        }
        .buttonStyle(.plain)
        .accessibilityLabel(L10n.text("note_source.accessibility.open", source.platformName, source.title))
    }

    private var sourceBadge: some View {
        ZStack {
            Circle()
                .fill(badgeColor)
                .frame(width: compact ? 30 : 34, height: compact ? 30 : 34)

            if let initials = badgeInitials {
                Text(initials)
                    .font(.system(size: compact ? 9 : 10, weight: .bold))
                    .foregroundColor(.white)
                    .lineLimit(1)
                    .minimumScaleFactor(0.6)
                    .padding(.horizontal, 3)
            } else {
                Image(systemName: badgeSystemImage)
                    .font(.system(size: compact ? 14 : 15, weight: .semibold))
                    .foregroundColor(.white)
            }
        }
        .accessibilityHidden(true)
    }

    private var badgeInitials: String? {
        switch source.platformID {
        case "xiaohongshu":
            return "小红书"
        case "youtube":
            return "YT"
        case "tiktok":
            return "TT"
        case "instagram":
            return "IG"
        case "threads":
            return "TH"
        case "reddit":
            return "RD"
        case "pinterest":
            return "PI"
        case "linkedin":
            return "IN"
        case "facebook":
            return "FB"
        case "vimeo":
            return "VI"
        case "twitch":
            return "TW"
        case "product_hunt":
            return "PH"
        case "hacker_news":
            return "HN"
        case "bilibili":
            return "B"
        default:
            return nil
        }
    }

    private var badgeSystemImage: String {
        switch source.platformID {
        case "x":
            return "xmark"
        case "spotify", "apple_podcasts":
            return "waveform"
        default:
            return "link"
        }
    }

    private var badgeColor: Color {
        switch source.platformID {
        case "xiaohongshu":
            return Color(red: 0.96, green: 0.18, blue: 0.25)
        case "youtube":
            return Color(red: 1.0, green: 0.0, blue: 0.0)
        case "tiktok":
            return Color(red: 0.05, green: 0.06, blue: 0.08)
        case "instagram":
            return Color(red: 0.82, green: 0.25, blue: 0.55)
        case "threads":
            return Color(red: 0.08, green: 0.08, blue: 0.08)
        case "x":
            return Color(red: 0.08, green: 0.09, blue: 0.10)
        case "reddit":
            return Color(red: 1.0, green: 0.27, blue: 0.0)
        case "pinterest":
            return Color(red: 0.9, green: 0.0, blue: 0.12)
        case "linkedin":
            return Color(red: 0.0, green: 0.47, blue: 0.71)
        case "facebook":
            return Color(red: 0.09, green: 0.47, blue: 0.95)
        case "vimeo":
            return Color(red: 0.1, green: 0.69, blue: 0.93)
        case "twitch":
            return Color(red: 0.39, green: 0.25, blue: 0.65)
        case "product_hunt":
            return Color(red: 0.85, green: 0.27, blue: 0.15)
        case "hacker_news":
            return Color(red: 1.0, green: 0.4, blue: 0.0)
        case "bilibili":
            return Color(red: 0.0, green: 0.63, blue: 0.86)
        case "spotify":
            return Color(red: 0.12, green: 0.73, blue: 0.33)
        default:
            return .accentPrimary
        }
    }
}

File: android/app/src/main/java/com/sponteoai/chillscript/ui/source/NoteSourceCard.kt
package com.sponteoai.chillscript.ui.source

import androidx.annotation.DrawableRes
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.OpenInNew
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material.icons.outlined.GraphicEq
import androidx.compose.material.icons.outlined.Link
import androidx.compose.material.icons.outlined.NorthEast
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.onClick
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.sponteoai.chillscript.R
import com.sponteoai.chillscript.domain.NoteSourceMetadata
import com.sponteoai.chillscript.domain.authorDisplayName
import com.sponteoai.chillscript.ui.theme.ChillColors

/** Android rendering of the current iOS NoteSourceCard. */
@Composable
fun NoteSourceCard(
    source: NoteSourceMetadata,
    compact: Boolean = false,
    onOpen: () -> Unit,
) {
    val author = source.authorDisplayName?.takeIf { it.isNotBlank() }
    val showsDescription = source.title.trim().let { title ->
        title.isNotEmpty() &&
            !title.equals(source.platformName, ignoreCase = true) &&
            !title.equals(source.host, ignoreCase = true)
    }
    val accessibilityLabel = stringResource(
        R.string.note_source_accessibility_open,
        source.platformName,
        source.title,
    )
    val compactShape = RoundedCornerShape(14.dp)

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .then(if (compact) Modifier.clip(compactShape).background(ChillColors.BackgroundSecondary) else Modifier)
            .clickable(onClick = onOpen)
            .clearAndSetSemantics {
                role = Role.Button
                contentDescription = listOfNotNull(accessibilityLabel, author).joinToString(". ")
                onClick {
                    onOpen()
                    true
                }
            }
            .padding(
                horizontal = if (compact) 10.dp else 0.dp,
                vertical = if (compact) 9.dp else 10.dp,
            ),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(if (compact) 10.dp else 14.dp),
    ) {
        SourceBadge(source.platformId, compact)

        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(if (compact) 1.dp else 4.dp),
        ) {
            if (compact) {
                if (author != null) {
                    Text(
                        text = author,
                        color = ChillColors.TextSub,
                        fontSize = 12.sp,
                        lineHeight = 15.sp,
                        fontWeight = FontWeight.SemiBold,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
                if (showsDescription) {
                    Text(
                        text = source.title,
                        color = ChillColors.TextMain,
                        fontSize = 13.sp,
                        lineHeight = 17.sp,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            } else {
                if (showsDescription) {
                    Text(
                        text = source.title,
                        color = ChillColors.TextMain,
                        fontSize = 17.sp,
                        lineHeight = 22.sp,
                        fontWeight = FontWeight.SemiBold,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
                Text(
                    text = if (author == null) source.platformName else "$author · ${source.platformName}",
                    color = ChillColors.TextSub,
                    fontSize = 15.sp,
                    lineHeight = 20.sp,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }

        Spacer(Modifier.size(8.dp))
        Icon(
            imageVector = if (compact) Icons.Outlined.NorthEast else Icons.AutoMirrored.Outlined.OpenInNew,
            contentDescription = null,
            tint = ChillColors.TextSub,
            modifier = Modifier.size(if (compact) 12.dp else 18.dp),
        )
    }
}

@Composable
private fun SourceBadge(platformId: String, compact: Boolean) {
    val badgeSize = if (compact) 30.dp else 38.dp
    val markSize = if (compact) 17.dp else 21.dp
    Box(
        modifier = Modifier.size(badgeSize).background(platformColor(platformId), CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        val brand = platformBrandDrawable(platformId)
        val initials = platformInitials(platformId)
        when {
            brand != null -> Icon(
                painter = painterResource(brand),
                contentDescription = null,
                tint = Color.White,
                modifier = Modifier.size(markSize),
            )
            initials != null -> Text(
                text = initials,
                color = Color.White,
                fontSize = if (compact) 9.sp else 10.sp,
                fontWeight = FontWeight.Bold,
                maxLines = 1,
            )
            else -> Icon(
                imageVector = platformFallbackIcon(platformId),
                contentDescription = null,
                tint = Color.White,
                modifier = Modifier.size(if (compact) 14.dp else 17.dp),
            )
        }
    }
}

@DrawableRes
private fun platformBrandDrawable(id: String): Int? = when (id) {
    "youtube" -> R.drawable.source_brand_youtube
    "tiktok" -> R.drawable.source_brand_tiktok
    "instagram" -> R.drawable.source_brand_instagram
    else -> null
}

private fun platformInitials(id: String): String? = when (id) {
    "xiaohongshu" -> "小红书"
    "threads" -> "TH"
    "reddit" -> "RD"
    "pinterest" -> "PI"
    "linkedin" -> "IN"
    "facebook" -> "FB"
    "vimeo" -> "VI"
    "twitch" -> "TW"
    "product_hunt" -> "PH"
    "hacker_news" -> "HN"
    "bilibili" -> "B"
    else -> null
}

private fun platformFallbackIcon(id: String): ImageVector = when (id) {
    "x" -> Icons.Outlined.Close
    "spotify", "apple_podcasts" -> Icons.Outlined.GraphicEq
    else -> Icons.Outlined.Link
}

private fun platformColor(id: String): Color = when (id) {
    "xiaohongshu" -> Color(0xFFF52E40)
    "youtube" -> Color(0xFFFF0000)
    "tiktok" -> Color(0xFF0D0F14)
    "instagram" -> Color(0xFFD13F8C)
    "threads", "x" -> Color(0xFF141417)
    "reddit" -> Color(0xFFFF4500)
    "pinterest" -> Color(0xFFE6001F)
    "linkedin" -> Color(0xFF0078B5)
    "facebook" -> Color(0xFF1778F2)
    "vimeo" -> Color(0xFF1AB0ED)
    "twitch" -> Color(0xFF6441A5)
    "product_hunt" -> Color(0xFFD94526)
    "hacker_news" -> Color(0xFFFF6600)
    "bilibili" -> Color(0xFF00A1DB)
    "spotify" -> Color(0xFF1FBA54)
    else -> ChillColors.BrandTeal
}
